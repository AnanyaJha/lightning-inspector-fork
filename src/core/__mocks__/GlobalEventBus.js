/**
 * Mocks the GlobalEventBus to fire publish and subscribe calls right in the context, vs using a postMessage in
 * the parent window.
 */

const PUBLISH_KEY = "AuraInspector:publish";

export default class GlobalEventBus {

    /**
     * Broadcast a message to a listener at any level in the inspector. Including, the InjectedScript, the ContentScript or the DevToolsPanel object.
     *
     * @param  {String} key MessageID to broadcast.
     * @param  {Object} data any type of data to pass to the subscribe method.
     */
    static publish(key, data) {
        if(!key) { return; }
        // const PUBLISH_BATCH_KEY = "AuraInspector:publishbatch";
        // const BOOTSTRAP_KEY = "AuraInspector:bootstrap";

        const subscribers = GlobalEventBus.subscribers;
        const jsonData = JSON.stringify(data);
        this.callSubscribers(key, data);
    }

    /**
     * Listen for a published message through the system.
     *
     * @param  {String} key Unique MessageId that would be broadcast through the system.
     * @param  {Function} callback function to be executed when the message is published.
     */
    static subscribe(key, callback) {
        const subscribers = GlobalEventBus.subscribers;
        if(!subscribers.has(key)) {
            subscribers.set(key, []);
        }

        subscribers.get(key).push(callback);
    }

    /**
     * Call everyone who is listening for this subscriber.
     */
    static callSubscribers(key, data) {
        const subscribers = GlobalEventBus.subscribers;

        if(subscribers.has(key)) {
            subscribers.get(key).forEach(function(callback){
                try {
                    callback(data);
                } catch(e) {
                    console.error("Key: ", key, " resulted in error ", e);
                }
            });
        }
    }
}

GlobalEventBus.subscribers = new Map();