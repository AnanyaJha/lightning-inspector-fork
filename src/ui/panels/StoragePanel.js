import React from "react";
import PropTypes from 'prop-types';

import BrowserApi from "../../aura/viewer/BrowserApi.js";
import GlobalEventBus from "../../core/GlobalEventBus.js";

import StorageCard from "../components/StorageCard.js";
import { Button, ButtonIcon, StatefulButton, ExpandableSection, Box } from "react-lds";

export default class StoragePanel extends React.Component {
    static childContextTypes = {
        assetBasePath: PropTypes.string
    };
    
    constructor(props) {
        super(props);

        this.state = {
            storages: new Map()
        };

        this.handleRefreshButtonClick = this.handleRefreshButtonClick.bind(this);
        this.handleDeleteButtonClick = this.handleDeleteButtonClick.bind(this);

        GlobalEventBus.subscribe("AuraInspector:UpdatedStorageData", this.handleStorageDataUpdate.bind(this));
        GlobalEventBus.subscribe("AuraInspector:ModifiedStorageData", this.handleModifiedStorageData.bind(this));
        GlobalEventBus.subscribe("AuraInspector:DeletedStorageData", this.handleDeletedStorageData.bind(this));
    }

    getChildContext() {
        return { 
            'assetBasePath': '/dist/slds/'
        };
    }

    /**
     * Gets the stores state, and sends the udpated data via PostMessage to the listeners.
     */
    async getStoresData(stores) {
        const count = stores && stores.length || 0;

        if(!count) {
            return;
        }

        const commands = [];
        for (let c=0;c<count;c++) {
            let store = stores[c];
            commands.push(`
                    var output = "o_${store}";
                    if(!data["o_${store}"]) {data["o_${store}"] = {}};
                    data["i_${store}"] = $A.storageService.getStorage('${store}');

                    // sync
                    data["o_${store}"].name = data["i_${store}"].getName();
                    data["o_${store}"].maxSize = data["i_${store}"].getMaxSize();
                    data["o_${store}"].version = data["i_${store}"].getVersion();

                    // async
                    data["i_${store}"].getSize()
                        .then(function(size) { if(!data["o_${store}"]) {data["o_${store}"] = {}} data["o_${store}"].size = size; }, function(err) { if(!data["o_${store}"]) {data["o_${store}"] = {}} data["o_${store}"].size = JSON.stringify(err); })
                        .then(function() { return data["i_${store}"].getAll(); })
                        .then(function(all) { if(!data["o_${store}"]) {data["o_${store}"] = {}} data["o_${store}"].all = all; }, function(err) { data["o_${store}"].all = JSON.stringify(err); })
                        // last then() is to post the results to aura inspector
                        .then(function() { window.postMessage({action:'AuraInspector:publish', key: 'AuraInspector:UpdatedStorageData', data:{ id:'${store}', data: JSON.stringify(data["o_${store}"])} }, window.location.href); });

            `);
        }

        const builtCommands = commands.join('\n');
        const command = `
            {
                var data = {};
                ${builtCommands};
                data;
            }
        `;
    
        try {
            await BrowserApi.eval(command);
        } catch(exceptionInfo) {
            console.error(exceptionInfo);
        }
    }

    async getStoresList() {
        // must collect the store names before doing the cache update
        const command = "Object.keys($A.storageService.getStorages());";
        const response = await BrowserApi.eval(command);
        
        if(!response || !response.map) { return; }

        // Replace ' \' '  to '_' so that template strings doesn't break.
        return response.map(function(name){ return name.replace('\'', '_'); });
    }

    async handleRefreshButtonClick() {
        this.update();
    }

    handleDeleteButtonClick() {
        const storageSelected = this.refs["storages-list"].selectedOptions;
        if(storageSelected.length === 0 || !storageSelected[0].value) {
            return;
        }

        const storageId = storageSelected[0].value;
        const command = `
            $A.storageService.deleteStorage('${storageId}').then(() => {
                window[Symbol.for('AuraDevTools')].Inspector.publish('AuraInspector:DeletedStorageData', '${storageId}');
            }).catch((error) => {
                window[Symbol.for('AuraDevTools')].Inspector.reportError(error);
            });
        `;

        return BrowserApi.eval(command);
    }

    async update() {
        const stores = await this.getStoresList();
        this.getStoresData(stores);
    }

    handleStorageDataUpdate(storageData) {
        const storageId = storageData.id;
        const storageContents = storageData.data;

        const storages = this.state.storages;
        storages.set(storageId, storageContents);

        this.setState({
            storages: storages
        });
    }

    handleModifiedStorageData(updatedStoreId) {
        this.getStoresData([updatedStoreId]);
    }

    handleDeletedStorageData(deletedStoreId) {
        this.state.storages.delete(deletedStoreId);

        this.setState({
            storages: this.state.storages
        });
    }

    getStorageElements() {
        const elements = [];
        const isExpanded = this.state.storages.size < 5;

        this.state.storages.forEach((storageContents, storageId) => {
            const preFormatted = JSON.parse(storageContents);
            const formatted = {};

            formatted.Adapter = preFormatted.name;
            formatted.sizeEstimate = "";

            if (preFormatted.size !== undefined) {
                const sizeAsPercent = (preFormatted.size / preFormatted.maxSize * 100).toFixed(0);
                formatted.sizeEstimate = preFormatted.size.toFixed(1) + " KB (" + sizeAsPercent + "% of " + preFormatted.maxSize.toFixed(0) + " KB)";
            }

            formatted.Version = preFormatted.version;
            formatted.Items = preFormatted.all || {};

            elements.push(<StorageCard key={storageId} storageId={storageId} isExpanded={isExpanded} rawContents={storageContents} contents={formatted}/>);
        });

        if(!elements.length) {
            return (<Box theme="warning" className="slds-m-around_large">{BrowserApi.getLabel("storage_empty")}</Box>);
        }

        return elements;
    }

    getStorageListElements() {
        const elements = [];
        this.state.storages.forEach((storageContents, storageId) => {
            elements.push(<option key={storageId} value={storageId}>{storageId}</option>);
        });
        return elements;
    }

    componentDidMount() {
        this.update();
    }

    render() {
        const labels = {
            "refresh": BrowserApi.getLabel("menu_refresh"),
            "kb": BrowserApi.getLabel("kb"), // Kilobytes not knowledgebase
            "delete": BrowserApi.getLabel("menu_delete"),
            "select_label": BrowserApi.getLabel("storage_selecttodestroy"),
            "delete_tooltip": BrowserApi.getLabel("storage_delete_tooltip")
        };

        const Fragment = React.Fragment;

        return (
            <Fragment>
                <menu type="toolbar" className="slds-col">
                    <li className="slds-p-right_xx-small">
                        <Button className="slds-p-around_xx-small" onClick={this.handleRefreshButtonClick} title={labels.refresh} tooltip={labels.refresh}>
                          <ButtonIcon sprite="utility" icon="refresh" />
                        </Button>
                    </li>
                    <li className="divider"></li>
                    <li className="slds-p-left_xx-small slds-p-right_xx-small">
                        <select className="slds-select" ref="storages-list">
                            <option>{labels.select_label}</option>
                            <option>--------------------------</option>
                            {
                                this.getStorageListElements()                                
                            }
                        </select>
                    </li>
                    <li>
                        <Button title={labels.delete} flavor={["destructive"]} tooltip={labels.delete_tooltip} onClick={this.handleDeleteButtonClick} />
                    </li>
                </menu>
                <div className="slds-col slds-scrollable_y slds-p-bottom_large inspector-tab-body-content">
                    {
                        this.getStorageElements()
                    }
                </div>
            </Fragment>
            );
    }

}