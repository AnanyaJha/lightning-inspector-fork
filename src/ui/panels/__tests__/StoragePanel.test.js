import React from 'react';
import StoragePanel from "../StoragePanel.js";
import renderer from "react-test-renderer";
import GlobalEventBus from "../../../core/GlobalEventBus.js";


import { Box } from "react-lds";

jest.mock("../../../aura/viewer/BrowserApi.js");
jest.mock("../../../core/GlobalEventBus.js");

/**
    StoragePanel Test Plan

   [x] 1. When you provide an empty data set, a warning message is shown there are no storages on the page.
   [x] 2. When you provide 1 storage, 1 storage shows on the page.
   [x] 3. When you provice X storages, x storages show up on the page.
   [ ] 4. Test "AuraInspector:UpdatedStorageData", "AuraInspector:ModifiedStorageData", "AuraInspector:DeletedStorageData"
   [ ] 5. Test Error condition fetching the data.
   [ ] 6. Calling "AuraInspector:ModifiedStorageData" refetches data for that data store.

    ---
    Notes:
    * Use Mock data for the storages.
    * Can we do some automation for verifying that the data we got from Aura is what we expect?
     - Yes, but we'd need an aura server and webdriver automation.
     - We'll need a CI that can start up an Aura Server, provide our own namespace of components, and run integration tests against them.  
*/


test("Snapshot of StoragePanel", () => {
    const component = renderer.create(<StoragePanel />);
    
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("Warning Message appears for no storages", () => {
    const component = renderer.create(<StoragePanel />);

    const testInstance = component.root;

    expect(testInstance.findAllByType(Box)).toHaveLength(1);

});


test("Warning message does not appear when storages exist", () => {
    const component = renderer.create(<StoragePanel />);

    GlobalEventBus.publish("AuraInspector:UpdatedStorageData", new StorageData("__test__", "__testname__", 1));

    const testInstance = component.root;

    expect(testInstance.findAllByType(Box)).toHaveLength(0);
});


test("One storage shows", () => {
    const component = renderer.create(<StoragePanel />);

    GlobalEventBus.publish("AuraInspector:UpdatedStorageData", new StorageData("__test__", "__testname__", 1, 2000, 10000));

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("Add then remove storages shows empty message", () => {
    const component = renderer.create(<StoragePanel />);
    const storageData = new StorageData("__test__", "__testname__", 1);

    GlobalEventBus.publish("AuraInspector:UpdatedStorageData", storageData);
    GlobalEventBus.publish("AuraInspector:DeletedStorageData", storageData);
    const testInstance = component.root;

    expect(testInstance.findAllByType(Box)).toHaveLength(0);
});

test("Calling AuraInspector:ModifiedStorageData updates data for that data store", () => {
    const storageId = "__storageId__";
    const component = renderer.create(<StoragePanel />);
    const mockFunc = jest.fn();
    component.getInstance().getStoresData=mockFunc;

    GlobalEventBus.publish("AuraInspector:ModifiedStorageData", storageId);
    
    expect(mockFunc.mock.calls.length).toBe(1);
    expect(mockFunc.mock.calls[0][0][0]).toBe(storageId);
});



class StorageData {
    constructor(id, name, version, size, maxSize) {
        this.id = id;
        this.data = {
            "name": name,
            "version": version
        };

        if(size !== undefined) {
            this.data.size = size;
        }

        if(maxSize !== undefined) {
            this.data.maxSize = maxSize;
        }

        this.data = JSON.stringify(this.data);
    }
}