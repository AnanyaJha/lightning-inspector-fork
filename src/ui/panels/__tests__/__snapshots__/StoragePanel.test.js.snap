// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`One storage shows 1`] = `
Array [
  <menu
    className="slds-col"
    type="toolbar"
  >
    <li
      className="slds-p-right_xx-small"
    >
      <button
        className="slds-button  slds-p-around_xx-small"
        disabled={false}
        href={null}
        onClick={[Function]}
        title="Refresh"
        value={null}
      >
        <svg
          aria-hidden="true"
          className="slds-button__icon"
        >
          <use
            xlinkHref="/dist/slds/assets/icons/utility-sprite/svg/symbols.svg#refresh"
          />
        </svg>
      </button>
    </li>
    <li
      className="divider"
    />
    <li
      className="slds-p-left_xx-small slds-p-right_xx-small"
    >
      <select
        className="slds-select"
      >
        <option>
          Select store to destroy
        </option>
        <option>
          --------------------------
        </option>
        <option
          value="__test__"
        >
          __test__
        </option>
      </select>
    </li>
    <li>
      <button
        className="slds-button slds-button_destructive"
        disabled={false}
        href={null}
        onClick={[Function]}
        title="Delete Storage"
        value={null}
      >
        Delete
      </button>
    </li>
  </menu>,
  <div
    className="slds-col slds-scrollable_y slds-p-bottom_large inspector-tab-body-content"
  >
    <div
      className="slds-section slds-is-open slds-p-horizontal_x-small"
    >
      <h3
        className="slds-section__title"
      >
        <button
          aria-controls="__test__"
          aria-expanded={true}
          className="slds-button  slds-section__title-action"
          disabled={false}
          href={null}
          onClick={[Function]}
          title={null}
          value={null}
        >
          <svg
            aria-hidden="true"
            className="slds-button__icon slds-button__icon_left"
          >
            <use
              xlinkHref="/dist/slds/assets/icons/utility-sprite/svg/symbols.svg#chevrondown"
            />
          </svg>
          <span
            className="slds-truncate"
            title="__test__"
          >
            __test__
          </span>
        </button>
      </h3>
      <div
        aria-hidden={true}
        className="slds-section__content"
        id="__test__"
      >
        <aurainspector-json
          expandTo="2"
        >
          {"Adapter":"__testname__","sizeEstimate":"2000.0 KB (20% of 10000 KB)","Version":1,"Items":{}}
        </aurainspector-json>
      </div>
    </div>
  </div>,
]
`;

exports[`Snapshot of StoragePanel 1`] = `
Array [
  <menu
    className="slds-col"
    type="toolbar"
  >
    <li
      className="slds-p-right_xx-small"
    >
      <button
        className="slds-button  slds-p-around_xx-small"
        disabled={false}
        href={null}
        onClick={[Function]}
        title="Refresh"
        value={null}
      >
        <svg
          aria-hidden="true"
          className="slds-button__icon"
        >
          <use
            xlinkHref="/dist/slds/assets/icons/utility-sprite/svg/symbols.svg#refresh"
          />
        </svg>
      </button>
    </li>
    <li
      className="divider"
    />
    <li
      className="slds-p-left_xx-small slds-p-right_xx-small"
    >
      <select
        className="slds-select"
      >
        <option>
          Select store to destroy
        </option>
        <option>
          --------------------------
        </option>
      </select>
    </li>
    <li>
      <button
        className="slds-button slds-button_destructive"
        disabled={false}
        href={null}
        onClick={[Function]}
        title="Delete Storage"
        value={null}
      >
        Delete
      </button>
    </li>
  </menu>,
  <div
    className="slds-col slds-scrollable_y slds-p-bottom_large inspector-tab-body-content"
  >
    <div
      className="slds-box slds-theme_warning slds-m-around_large"
    >
      No Aura Storages located on the page
    </div>
  </div>,
]
`;
